### PR TITLE
[Tools] Added tool to check for data consistency in flags and instrument tables

### PR DESCRIPTION
**Pull request by AnyhowStep**
The code does these checks:

+ Tables specified in `test_names` exist
+ Orphan flag rows
+ Orphan instrument rows
+ Duplicate instrument flags in a session

Run with: `php tools/instrument_flag_consistency_check.php`

Screenshots:

image
image

### CODE REVIEW

**Comment by Jkat** [comment]
Can you do one check that goes through every commentID in the flag table and reconstructs the commentID(minus the unix timestamp) and looks for discrepancies?

**Comment by Jkat** [comment]
Given commentID, get test_name and sessionID.
Given test_name, get the test ID
Given sessionID, get CandID, SubprojectID
Given CandID, get PSCID
Concat CandID . PSCID . sessionID . SubprojectID . TestID
Check against commentID(without unix timestamp)

**Comment by driusan** [comment]
How long does this script take? Is it fast enough to be feasible to put it in a front-end module instead of a tools script that requires shell access to LORIS?

Data discrepancies are something that study coordinators are concerned about, so it makes more sense to do it somewhere that's more accessible for them than the shell.

**Comment by ridz1208** [comment]
@driusan at the same time some of these errors could be caused by mismanipulation of the data on the DCC end... maybe we do not want study coordinators involved in this. especially if there are more than a dozen... then start considering permissions and site restrictions 

I think this should be a bash script personally

File: tools/instrument_flag_consistency_check.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            return str_pad($str, $len);
+        }
+    }
+    
+    require_once __DIR__ . "/../vendor/autoload.php";
+    $client = new NDB_Client();
+    $client->makeCommandLine();
+    $client->initialize(__DIR__ . "/../project/config.xml");
+    $DB = Database::singleton();
+
```
Comment (ridz1208) [inline tools/instrument_flag_consistency_check.php]: @AnyhowStep please use generic includes.

File: tools/instrument_flag_consistency_check.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                Test_name = :check AND
+                {$check}.CommentID IS NULL
+        ", array(
+            "check" => $check
+        ));
+        
+        $flag_str = pad_or_trunc($check, 5) . " flag";
+        foreach ($orphan_flag_arr as $orphan_flag) {
+            $comment_id = $orphan_flag["CommentID"];
+            echo "Warning, orphan {$flag_str}:{$comment_id}\n";
```
Comment (ridz1208) [inline tools/instrument_flag_consistency_check.php]: @AnyhowStep make a distinction between flag orphans and instrument orphans (line 82)

File: tools/instrument_flag_consistency_check.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            ON
+                {$check}.CommentID = flag.CommentID
+            WHERE
+                flag.CommentID IS NULL
+        ", array());
+        
+        $check_str = pad_or_trunc($check, 10);
+        foreach ($orphan_instrument_arr as $orphan_instrument) {
+            $comment_id = $orphan_instrument["CommentID"];
+            echo "Warning, orphan {$check_str}:{$comment_id}\n";
```
Comment (ridz1208) [inline tools/instrument_flag_consistency_check.php]: printout truncating full name 

`Warning, orphan Clin. flag:918988PKD56742213771479426256`

File: tools/instrument_flag_consistency_check.php (hunk 2 above)
Comment (AnyhowStep) [inline tools/instrument_flag_consistency_check.php]: There is a distinction. One says flag, the other does not. I'm not sure what's required?

File: tools/instrument_flag_consistency_check.php (hunk 3 above)
Comment (AnyhowStep) [inline tools/instrument_flag_consistency_check.php]: It is truncating the full name to 10 characters here.

File: tools/instrument_flag_consistency_check.php (hunk 1 above)
Comment (AnyhowStep) [inline tools/instrument_flag_consistency_check.php]: Which tools are using the `generic_includes.php` file? I looked over a few of them and saw them doing exactly this.

**Comment by AnyhowStep** [comment]
@ridz1208 
What kind of comments would you like to see?
The process is broken down and separated into 3 parts, each part has an echo stating exactly what each part will do.

Each part one or two queries and a for-loop.

**Comment by AnyhowStep** [comment]
@driusan I'm actually kind of tempted to either paginate the results or something. At least not load all results into memory at once. I remember trying to load 500k results on my VM made it rebel until I added sufficient swap space.

I'm thinking worst case scenario type thing here. All of their data for a particular instrument is bad. And they have 500k results. Loading it will just cause the web server process to crash unless they increase the space allocated for PHP/web server and add enough swap.

It's as "fast" as loading and printing all problems can be. It could be modified to be broken into a few parts,
1) Just count the total number of discrepancies per instrument.
2) Allow user to click on a particular instrument to view a paginated list of discrepancies. (orphan/duplicate)

This will allow it to run "faster" because it's not doing everything in one operation, I guess. But I don't know what's desired.

File: tools/instrument_flag_consistency_check.php (hunk 3 above)
Comment (ridz1208) [inline tools/instrument_flag_consistency_check.php]: dont

File: tools/instrument_flag_consistency_check.php (hunk 2 above)
Comment (ridz1208) [inline tools/instrument_flag_consistency_check.php]: I dont see it in the print out, not clearly anyways

File: tools/instrument_flag_consistency_check.php (hunk 1 above)
Comment (ridz1208) [inline tools/instrument_flag_consistency_check.php]: all the new ones

**Comment by johnsaigle** [comment]
It might be useful to present this in a LORIS meeting so others can comment on what they would like to see/what's useful. Since this PR is two months old and the discussion is pretty open-ended, I think it could benefit from offline communication.

**Comment by driusan** [comment]
This PR needs someone to adopt it. Adding the "Blocked" tag and moving to 17.2 since it needs someone to pick it up and I don't foresee that happening for 17.1..

**Comment by ridz1208** [comment]
@driusan I'd like to take a stab at it OR it could be merged and then progressively made better...

**Comment by driusan** [comment]
@ridz1208 I don't think there's any urgency for it that we should just merge it. Can you close this PR and send a new one to 17.2 so that you get the notifications about it?

**Comment by ridz1208** [comment]
add check for examiners with no sites associated
add check for users with no site associated

**Comment by ridz1208** [comment]
add https://github.com/aces/CCNA/pull/1605

**Comment by ridz1208** [comment]
this Pr contains valuable work that will be extended in a near future.

Due to the fast changing codebase and the major work towards a more consistent database. This pr will be re-worked as per redmine ticket
https://redmine.cbrain.mcgill.ca/issues/13717

**Pull request closed without merging**